README
NAME
    Plack::App::RDF::Files - serve RDF data from files

SYNOPSIS
    Create and run a Linked Open Data server in one line:

        plackup -e 'use Plack::App::RDF::Files "app"; app(base_dir=>"/path/to/rdf")'

    In more detail, create a file "app.psgi":

        use Plack::App::RDF::Files;
        Plack::App::RDF::Files->new(
            base_dir => '/path/to/rdf/',       # mandatory
            base_uri => 'http://example.org/'  # optional
        )->to_app;

    Run it as web application by calling "plackup". Request URLs are then
    mapped to URIs and directories to return data from RDF files as
    following:

        http://localhost:5000/foo  =>  http://example.org/foo
                                             /path/to/rdf/foo/
                                             /path/to/rdf/foo/*.(nt|ttl|rdfxml)
        http://localhost:5000/x/y  =>  http://example.org/x/y
                                             /path/to/rdf/x/y/
                                             /path/to/rdf/x/y/*.(nt|ttl|rdfxml)

    In short, each subdirectory corresponds to an RDF resource.

DESCRIPTION
    This PSGI application serves RDF from files. Each accessible RDF
    resource corresponds to a (sub)directory, located in a common based
    directory. All RDF files in a directory are merged and returned as RDF
    graph. If no RDF data was found in an existing subdirectory, an
    axiomatic triple is returned:

        $REQUEST_URI <a <http://www.w3.org/2000/01/rdf-schema#Resource> .

    Requesting the base directory, however will result in a HTTP 404 error
    unless option "index_property" is enabled.

    HTTP HEAD and conditional GET requests are supported by ETag and
    Last-Modified-Headers (see Plack::Middleware::ConditionalGET).

CONFIGURATION
    base_dir
        Mandatory base directory that all resource directories are located
        in.

    base_uri
        The base URI of all resources. If no base URI has been specified,
        the base URI is taken from the PSGI request.

    file_types
        An array of RDF file types, given as extensions to look for. Set to
        "['rdfxml','nt','ttl']" by default.

    index_property
        By default a HTTP 404 error is returned if one tries to access the
        base directory. Enable this option by setting it to 1 or to an URI,
        to also serve RDF data from the base directory. By default
        "http://www.w3.org/2000/01/rdf-schema#seeAlso" is used as index
        property, if enabled.

    path_map
        Optional code reference that maps a local part of an URI to a
        relative directory. Set to the identity mapping by default.

    namespaces
        Optional namespaces for serialization, passed to
        RDF::Trine::Serializer.

METHODS
  call( $env )
    Core method of the PSGI application.

    The following PSGI environment variables are read and/or set by the
    application.

    rdf.uri
        The requested URI as string or URI object.

    rdf.iterator
        The RDF::Trine::Iterator that will be used for serializing, if
        "psgi.streaming" is set. One can use this variable to catch the RDF
        data in another post-processing middleware.

    rdf.files
        An hash of source filenames, each with the number of triples (on
        success) as property "size", an error message as "error" if parsing
        failed, and the timestamp of last modification as "mtime". "size"
        and "error" may not be given before parsing, if "rdf.iterator" is
        set.

    negotiate.format
        RDF serialization format (See Plack::Middleware::Negotiate).
        Supported values are "ttl", "nt", "n3", "json", and "rdfxml".

    If an existing resource does not contain triples, the axiomatic triple
    "$uri rdf:type rdfs:Resource" is returned.

  files( $env )
    Get a list of RDF files (as hash reference) that will be read for a
    given request, given as PSGI environment.

    The requested URI is saved in field "rdf.uri" of the request
    environment. On success returns the base directory and a list of files,
    each mapped to its last modification time. Undef is returned if the
    request contained invalid characters (everything but "a-zA-Z0-9:.@/-"
    and the forbidden sequence "../" or a sequence starting with "/"), or if
    called with the base URI and "index_property" not enabled.

  headers( $files )
    Get a response headers object (as provided by Plack::Util::headers) with
    ETag and Last-Modified from a list of RDF files given as returned by the
    files method.

FUNCTIONS
  app( %options )
    This shortcut for "Plack::App::RDF::Files->new" can be exported on
    request to simplify one-liners.

SEE ALSO
    Use Plack::Middleware::Negotiate to add content negotiation based on an
    URL parameter and/or suffix.

    See RDF::LinkedData for a different module to serve RDF as linked data.
    See also RDF::Flow and RDF::Lazy for processing RDF data.

    See <http://foafpress.org/> for a similar approach in PHP.

COPYRIGHT AND LICENSE
    Copyright Jakob Voss, 2014-

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
# Changelog for Plack-App-RDF-Files

0.11  2014-12-12 12:02:37 CET
    - improved serialization, especially for Unicode characters
    - added 'app' function to simplify one-liners

0.10  2014-12-10 15:12:08 CET
    - fix non-streaming Unicode
    - migrated to Dist::Milla and Moo
    - major code cleanup and refactoring
    - add HTTP HEAD and conditional GET
    - fix feature index_property, remove include_index

0.02  2013-07-02
    - new method: files

0.01  2013-07-01
    - initial release


Makefile.PL

use strict;
use warnings;

use 5.014001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Serve RDF data from files",
  "AUTHOR" => "Jakob Vo\303\237 <voss\@gbv.de>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Plack-App-RDF-Files",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Plack::App::RDF::Files",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Digest::MD5" => 0,
    "Encode" => 0,
    "File::Spec::Functions" => 0,
    "HTTP::Date" => 0,
    "List::Util" => 0,
    "Plack::Component" => 0,
    "Plack::Request" => 0,
    "Plack::Util::Accessor" => 0,
    "RDF::Trine" => 0,
    "RDF::Trine::Iterator::Graph" => 0,
    "RDF::Trine::Model" => 0,
    "RDF::Trine::Parser" => 0,
    "RDF::Trine::Serializer" => 0,
    "Scalar::Util" => 0,
    "URI" => 0,
    "parent" => 0
  },
  "TEST_REQUIRES" => {
    "HTTP::Message::PSGI" => 0,
    "HTTP::Request" => 0,
    "HTTP::Request::Common" => 0,
    "Plack::Builder" => 0,
    "Plack::Test" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.02",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Plack/App/RDF/Files.pm
package Plack::App::RDF::Files;
{
  $Plack::App::RDF::Files::VERSION = '0.02';
}
#ABSTRACT: Serve RDF data from files

use v5.14;

use parent 'Plack::Component';
use Plack::Util::Accessor qw(
    base_dir base_uri file_types path_map
    include_index index_property namespaces
);

use Plack::Request;
use RDF::Trine qw(statement iri);
use RDF::Trine::Model;
use RDF::Trine::Parser;
use RDF::Trine::Serializer;
use RDF::Trine::Iterator::Graph;
use File::Spec::Functions qw(catfile catdir);
use URI;
use Scalar::Util qw(blessed reftype);
use Carp qw(croak);
use Digest::MD5 qw(md5_hex);
use HTTP::Date;
use List::Util qw(max);


our %FORMATS = (
    ttl     => 'Turtle',
    nt      => 'NTriples',
    n3      => 'Notation3',
    json    => 'RDFJSON',
    rdfxml  => 'RDFXML'
);

sub prepare_app {
    my $self = shift;
    return if $self->{prepared};

    die "missing base_dir" unless $self->base_dir and -d $self->base_dir;

    $self->base_uri( URI->new( $self->base_uri ) )
        if $self->base_uri;

    my $types = join '|', @{ $self->file_types // [qw(rdfxml nt ttl)] };
    $self->file_types( qr/^($types)/ );

    if ( $self->include_index ) {
        $self->index_property( 'http://www.w3.org/2000/01/rdf-schema#seeAlso' )
            unless defined $self->index_property;
        $self->index_property( iri( $self->index_property ) )
            if $self->index_property;
    }

    $self->path_map( sub { shift } ) unless $self->path_map;

    $self->{prepared} = 1;

    $self;
}


sub files {
    my $self = shift;

    my ($env, $req, $path);

    if (!reftype($_[0])) {                                      # $str
        return unless $self->base_uri and defined $_[0];
        # TODO: support full URIs via HTTP::Request
        $path = substr(shift,1);
        $env = { };
    } elsif (!blessed($_[0]) and reftype($_[0]) eq 'HASH') {    # $env
        $env  = shift;
        $req  = Plack::Request->new($env);
        $path = substr($req->path,1);
    } elsif (blessed($_[0]) and $_[0]->isa('Plack::Request')) { # $req
        $req  = shift;
        $env  = $req->env;
        $path = substr($req->path,1);
    } else {
        croak "expected PSGI request or string";
    }

    return if $path !~ /^[a-z0-9:\._@\/-]*$/i or $path =~ /\.\.\/|^\//;

    $env->{'rdf.uri'} = URI->new( ($self->base_uri // $req->base) . $path );

    return if $path eq '' and !$self->include_index;

    my $dir = catdir( $self->base_dir, $self->path_map->($path) );

    return unless -d $dir;
    return ($dir) unless -r $dir and opendir(my $dh, $dir);

    my $files = { };
    while( readdir $dh ) {
        next if $_ !~ /\.(\w+)$/ || $1 !~ $self->file_types;
        my $full = catfile( $dir, $_ );
        $files->{$_} = {
            full  => $full,
            size  => (stat($full))[7],
            mtime => (stat($full))[9],
        }
    }
    closedir $dh;

    return ( $dir => $files );
}


sub call {
    my ($self, $env) = @_;
    my $req = Plack::Request->new($env);

    return [ 405, [ 'Content-type' => 'text/plain' ], [ 'Method not allowed' ] ]
        unless (($req->method eq 'GET') || ($req->method eq 'HEAD'));

    my ($dir, $files) = $self->files( $req );

    if (!$files) {
        my $status  = 404;
        my $message = $req->env->{'rdf.uri'}
                    ? "Not found: " . $req->env->{'rdf.uri'} : "Not found";

        if ($dir and -d $dir) {
            $status = 404;
            $message =~ s/found/accesible/;
        }

        return [ $status, [ 'Content-type' => 'text/plain' ], [ $message ] ];
    }

    my $uri = $env->{'rdf.uri'};
    my @headers;

    # TODO: show example with Plack::Middleware::ConditionalGET

    my $md5 = md5_hex( map { values %{$_} } values %$files );
    push @headers, ETag => "W/\"$md5\"";

    my $lastmod = max map { $_->{mtime} } values %$files;
    push @headers, 'Last-Modified' => HTTP::Date::time2str($lastmod) if $lastmod;

    # TODO: HEAD method

    # parse RDF
    my $model = RDF::Trine::Model->new;
    my $triples = 0;
    foreach (keys %$files) { # TODO: parse sorted by modifcation time?
        my $file = $files->{$_};

        my $parser = RDF::Trine::Parser->guess_parser_by_filename( $file->{full} );
        eval {
            $parser->parse_file_into_model( $uri, $file->{full}, $model );
        };
        if ($@) {
            $file->{error} = $@;
        } else {
            $file->{triples} = $model->size - $triples;
            $triples = $model->size;
        }
    }
    $env->{'rdf.files'} = $files;

    my $iter = $model->as_stream;

    # add listing on base URI
    if ( $self->index_property and "$uri" eq ($self->base_uri // $req->base) ) {
        my $subject   = iri( $uri );
        my $predicate = $self->index_property;
        my @stms;
        opendir(my $dirhandle, $dir);
        foreach my $p (readdir $dirhandle) {
            next unless -d catdir( $dir, $p ) and $p !~ /^\.\.?$/;
            push @stms, statement(
                $subject,
                $predicate,
                RDF::Trine::Node::Resource->new( "$uri$p" )
            );
        }
        closedir $dirhandle;

        my $i2 = RDF::Trine::Iterator::Graph->new( \@stms );
        $iter = $iter->concat( $i2 );
    }

    # add axiomatic triple to empty graphs
    if ($iter->finished) {
        $iter = RDF::Trine::Iterator::Graph->new( [ statement(
            iri($uri),
            iri('http://www.w3.org/1999/02/22-rdf-syntax-ns#type'),
            iri('http://www.w3.org/2000/01/rdf-schema#Resource')
        ) ] );
    }

    # TODO: do not serialize at all on request

    # TODO: HTTP HEAD method

    # negotiate and serialize
    my ($ser, @h) = $self->negotiate( $env, $uri );
    push @headers, @h;

    if (!$ser) {
        $ser = RDF::Trine::Serializer->new( 'NTriples', base_uri => $uri );
        @headers = ('Content-type' => 'text/plain');
    }

    if ( $env->{'psgi.streaming'} ) {
        $env->{'rdf.iterator'} = $iter;
        return sub {
            my $responder = shift;
            # TODO: use IO::Handle::Iterator to serialize as last as possible
            my $rdf = $ser->serialize_iterator_to_string( $iter );
            use Encode; # must be bytes
            $rdf = encode("UTF8",$rdf);
            $responder->( [ 200, [ @headers ], [ $rdf ] ] );
        };
    } else {
        my $rdf = $ser->serialize_iterator_to_string( $iter );
        return [200, [ @headers ], [ $rdf ] ];
    }
}


sub negotiate {
    my ($self, $env) = @_;

    if ( $env->{'negotiate.format'} ) {
		# TODO: catch RDF::Trine::Error::SerializationError and log
        my $ser = eval {
            RDF::Trine::Serializer->new( 
                $FORMATS{$env->{'negotiate.format'}} // $env->{'negotiate.format'},
                base       => $env->{'rdflow.uri'},
                namespaces => ( $self->namespaces // { } ),
            )
        };
        # TODO: push @headers, Vary => 'Accept'; ?
        return ($ser);
    } else {
        my ($ctype, $ser) = RDF::Trine::Serializer->negotiate(
            request_headers => Plack::Request->new($env)->headers,
        );
        my @headers = ( 'Content-type' => $ctype, Vary => 'Accept' );
        return ($ser, @headers);
    }
}

1;

__END__

=pod

=head1 NAME

Plack::App::RDF::Files - Serve RDF data from files

=head1 VERSION

version 0.02

=head1 SYNOPSIS

    my $app = Plack::App::RDF::Files->new(
        base_dir => '/var/rdf/
    );

    # Requests URI            =>  RDF files
    # http://example.org/     =>  /path/to/rdf/*.(nt|ttl|rdfxml)
    # http://example.org/foo  =>  /path/to/rdf/foo/*.(nt|ttl|rdfxml)
    # http://example.org/x/y  =>  /path/to/rdf/x/y/*.(nt|ttl|rdfxml)

=head1 DESCRIPTION

This L<PSGI> application serves RDF from files. Each accessible RDF resource
corresponds to a (sub)directory, located in a common based directory. All RDF
files in a directory are merged and returned as RDF graph.

=head1 METHODS

=head2 files( $env | $req | $str )

Get a list of RDF files that will be read for a given request. The request can
be specified as L<PSGI> environment, as L<Plack::Request>, or as partial URI
that follows C<base_uri> (given as string). The requested URI is saved in field
C<rdf.uri> of the request environment.  On success returns the base directory
and a list of files, each mapped to its last modification time.  Undef is
returned if the request contained invalid characters (everything but
C<a-zA-Z0-9:.@-> and the forbidden sequence C<../>) or if the request equals ro
the base URI and C<include_index> was not enabled.

=head2 negotiate( $env )

This internal methods selects an RDF serializer based on the PSGI environment 
variable C<negotiate.format> (see L<Plack::Middleware::Negotiate>) or the 
C<negotiate> method of L<RDF::Trine::Serializer>. Returns first a 
L<RDF::Trine::Serializer> on success or C<undef> on error) and second a 
(possibly empty) list of HTTP response headers.

=head1 CONFIGURATION

=over 4

=item base_dir

Mandatory base directory that all resource directories are located in.

=item base_uri

The base URI of all resources. If no base URI has been specified, the
base URI is taken from the PSGI request.

=item file_types

An array of RDF file types (extensions) to look for. Set to
C<['rdfxml','nt','ttl']> by default.

=item include_index

By default a HTTP 404 error is returned if one tries to access the base
directory. Enable this option to also serve RDF data from this location.

=item index_property

RDF property to use for listing all resources connected to the base URI (if
<include_index> is enabled).  Set to C<rdfs:seeAlso> by default. Can be
disabled by setting a false value.

=item path_map

Optional code reference that maps a local part of an URI to a relative
directory. Set to the identity mapping by default.

=item namespaces

Optional namespaces for serialization, passed to L<RDF::Trine::Serializer>.

=back

=head1 PSGI environment variables

The following PSGI environment variables are relevant:

=over 4

=item rdf.uri

The requested URI

=item rdf.iterator

The L<RDF::Trine::Iterator> that will be used for serializing, if
C<psgi.streaming> is set. One can use this variable to catch the RDF
data in another post-processing middleware.

=item rdf.files

An hash of source filenames, each with the number of triples (on success)
as property C<size>, an error message as C<error> if parsing failed, and
the timestamp of last modification as C<mtime>. C<size> and C<error> may
not be given before parsing, if C<rdf.iterator> is set.

=back

=head1 LIMITATIONS

B<This module is an early developer release. Be warned!>

All resource URIs to be served must have a common URI prefix (such as
C<http://example.org/> above) and a local part that may be restricted to a
limited set of characters. For instance the character sequence C<../> is
not allowed.

=head1 NOTES

If an existing resource does not contain triples, the axiomatic triple
C<< ?uri rdf:type rdfs:Resource >> is returned.

To update the files, add a middleware that catches 404 and 202 responses.

=head1 TODO

VoID descriptions could be added, possibly with L<RDF::Generator::Void>.

=head1 SEE ALSO

Use L<Plack::Middleware::Negotiate> to add content negotiation based on
an URL parameter and/or suffix.

See L<RDF::LinkedData> for a different module to serve RDF as linked data.
See also L<RDF::Flow> and L<RDF::Lazy> for processing RDF data.

See L<http://foafpress.org/> for a similar approach in PHP.

=head1 AUTHOR

Jakob Voß <voss@gbv.de>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Jakob Voß.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
